zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/changes/StartPartitionScaleUpApplier.java

```diff
@@ -0,0 +1,150 @@
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.dynamic.config.changes;
+
+import io.camunda.zeebe.dynamic.config.changes.ConfigurationChangeAppliers.ClusterOperationApplier;
+import io.camunda.zeebe.dynamic.config.state.ClusterConfiguration;
+import io.camunda.zeebe.dynamic.config.state.RoutingState;
+import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling;
+import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling.ActivePartitions;
+import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling.AllPartitions;
+import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling.Mixed;
+import io.camunda.zeebe.protocol.Protocol;
+import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.Either.Left;
+import io.camunda.zeebe.util.Either.Right;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Starts the process of scaling up partitions. The routing state version is incremented and the
+ * request handling strategy is updated from {@link AllPartitions} to a {@link Mixed} one, with new
+ * partitions starting off as inactive.
+ *
+ * <p>We use the request handling strategy to detect whether scaling up is possible or not - we
+ * expect {@link AllPartitions} at the beginning, update it to a {@link Mixed} one and finish the
+ * scaling by going back to {@link AllPartitions}. If we find anything but {@link AllPartitions}, we
+ * assume that we are already scaling up and refuse the operation.
+ */
+final class StartPartitionScaleUpApplier implements ClusterOperationApplier {
+
+  private final PartitionScalingChangeExecutor partitionScalingChangeExecutor;
+  private final int desiredPartitionCount;
+
+  StartPartitionScaleUpApplier(
+      final PartitionScalingChangeExecutor partitionScalingChangeExecutor,
+      final int desiredPartitionCount) {
+    this.partitionScalingChangeExecutor = partitionScalingChangeExecutor;
+    this.desiredPartitionCount = desiredPartitionCount;
+  }
+
+  @Override
+  public Either<Exception, UnaryOperator<ClusterConfiguration>> init(
+      final ClusterConfiguration currentClusterConfiguration) {
+    if (desiredPartitionCount < Protocol.START_PARTITION_ID) {
+      return new Left<>(
+          new IllegalArgumentException(
+              "Desired partition count must be greater than %d"
+                  .formatted(Protocol.START_PARTITION_ID)));
+    }
+
+    if (desiredPartitionCount > Protocol.MAXIMUM_PARTITIONS) {
+      return new Left<>(
+          new IllegalArgumentException(
+              "Desired partition count must not exceed %d".formatted(Protocol.MAXIMUM_PARTITIONS)));
+    }
+
+    for (int partition = Protocol.START_PARTITION_ID;
+        partition <= desiredPartitionCount;
```

Comment:
:upside_down_face: Super minor, but this condition relies on the fact `START_PARTITION_ID = 1`. If it's not (e.g. 2), then this fails. So could be `partition < desiredPartitionCount + Protocol.START_PARTITION_ID` instead. Super nit picky, sorry, since when would we ever change the start partition ID? (never say never :smile:)

Reply:
Good catch. I can check `partition < Protocol.START_PARTITION_ID + desiredPartitionCount` instead :+1: